crates/core/src/live_socket/socket.rs

```diff
@@ -306,33 +211,187 @@ impl LiveSocket {
 
         let has_live_reload = live_reload_iframe.is_some();
 
-        debug!("iframe src: {live_reload_iframe:?}");
-
-        Ok(Self {
-            socket,
+        let out = Self {
+            connect_opts,
+            url: url.clone(),
+            format: format.to_string(),
             csrf_token,
             phx_id,
             phx_static,
             phx_session,
-            timeout,
-            url,
             dead_render,
             style_urls,
             has_live_reload,
             cookies,
-            format,
+        };
+
+        debug!("Session data successfully acquired {out:?}");
+
+        Ok(out)
+    }
+
+    /// reconstruct the live socket url from the session data
+    pub fn get_live_socket_url(&self) -> Result<Url, LiveSocketError> {
+        let websocket_scheme = match self.url.scheme() {
+            "https" => "wss",
+            "http" => "ws",
+            scheme => {
+                return Err(LiveSocketError::SchemeNotSupported {
+                    scheme: scheme.to_string(),
+                })
+            }
+        };
+
+        let port = self.url.port().map(|p| format!(":{p}")).unwrap_or_default();
+        let host = self.url.host_str().ok_or(LiveSocketError::NoHostInURL)?;
+
+        let mut websocket_url = Url::parse(&format!("{websocket_scheme}://{host}{port}"))?;
+
+        websocket_url
+            .query_pairs_mut()
+            .append_pair(LVN_VSN_KEY, LVN_VSN)
+            .append_pair(CSRF_KEY, &self.csrf_token)
+            .append_pair(MOUNT_KEY, "0")
+            .append_pair(FMT_KEY, &self.format);
+
+        websocket_url.set_path("/live/websocket");
+
+        debug!("websocket url: {websocket_url}");
+
+        Ok(websocket_url)
+    }
+}
+
+#[derive(uniffi::Object)]
+pub struct LiveSocket {
+    pub socket: Mutex<Arc<Socket>>,
```

Comment:
I'm confused. Why's socket a `Mutex<Arc<Socket>>`? Wouldn't the mutex be locked for the entire time that someone has a reference to the return of  `fn socket(&self)`?

Reply:
No, we have a really bizarre API underlying this which returns an immutable object inside and `Arc` because the socket was expected to be orchestrated over FFI.  So there is no way around the having the internal smart pointer. 

When you lock it and clone the `Arc` you get an immutable reference to the internal smart pointer without the mutex. The only thing that holds a reference to these sockets is the `LiveChannel` struct, so it's beholden upon the consumer to drop the old dead `LiveChannels`, which are now disconnected, to avoid a memory leak.